desub-current/src/decoder/decode_storage.rs

```diff
@@ -0,0 +1,330 @@
+	NameNotFound,
+}
+
+impl StorageDecoder {
+	/// Call [`super::decode_storage()`] to construct a [`StorageDecoder`].
+	pub(super) fn generate_from_metadata(metadata: &Metadata) -> StorageDecoder {
+		let entries_by_hashed_prefix = metadata
+			.storage_entries()
+			.enumerate()
+			.map(|(index, entries)| {
+				let prefix_hash = twox_128(entries.prefix().as_bytes());
+				let entry_by_hashed_name = entries
+					.entries()
+					.enumerate()
+					.map(|(entry_index, entry)| {
+						let name_hash = twox_128(entry.name.as_bytes());
+						(name_hash, entry_index)
+					})
+					.collect();
+				(prefix_hash, StorageEntries { index, entry_by_hashed_name })
+			})
+			.collect();
+
+		StorageDecoder { entries_by_hashed_prefix }
+	}
+
+	/// Decode the SCALE encoded bytes representing a storage entry lookup. These conceptually take the
+	/// form `twox_128(prefix) + twox_128(name) + rest`, where `rest` depends on the storage entry we're
+	/// keying into, and may be nothing at all for plain storage locations, or hashed keys to access maps.
+	pub fn decode_key<'m, 'b>(
+		&self,
+		metadata: &'m Metadata,
+		bytes: &mut &'b [u8],
+	) -> Result<StorageEntry<'m, 'b>, StorageDecodeError> {
+		// Step 1: reverse-lookup the hashed prefix+name part of the key, and get
+		// details about this storage location from our metadata.
+		let location = self.decode_prefix_and_name_to_location(bytes)?;
+		let storage_entry = metadata.storage_entry(location);
+
+		let prefix_str = storage_entry.prefix;
+		let name_str = &*storage_entry.metadata.name;
+
+		// Step 2: use the details held in metadata to infer what form the rest of
+		// the bytes should take, and decode accordingly.
+		match &storage_entry.metadata.ty {
+			FrameStorageEntryType::Plain(ty) => {
+				// No more work to do here; our storage entry is a plain prefix+name entry,
+				// so return the details of it:
+				Ok(StorageEntry {
+					prefix: prefix_str.into(),
+					name: name_str.into(),
+					ty: *ty,
+					details: StorageEntryType::Plain,
+				})
+			}
+			FrameStorageEntryType::Map { hashers, key, value } => {
+				// We'll consume some more data based on the hashers.
+				// First, get the type information that we need ready.
+				let key =
+					metadata.types().resolve(key.id()).ok_or_else(|| StorageDecodeError::TypeNotFound(key.id()))?;
+				let keys = storage_map_key_to_type_id_vec(metadata, key)?;
+				if keys.len() != hashers.len() {
+					return Err(StorageDecodeError::KeysAndHashersDontLineUp {
+						num_keys: keys.len(),
+						num_hashers: hashers.len(),
+					});
+				}
+
+				// Work through the hashers and type info we have to generate out output
+				// data, and consume bytes from the input cursor as we go.
+				let mut storage_keys = vec![];
+				for (hasher, ty) in hashers.iter().zip(keys) {
+					pub use frame_metadata::v14::StorageHasher as FrameStorageHasher;
+
+					// How many bytes will the hashed bit consume?
+					let initial_hash_bytes = match hasher {
```

Comment:
I wonder if it's possible to use `parity-scale-codec`'s `remaining_len` fn : https://docs.rs/parity-scale-codec/2.3.1/parity_scale_codec/trait.Input.html#tymethod.remaining_len

as opposed to this calculation here and the one below, would make it more dynamic maybe?

Reply:
I might be missing something, but I'm not sure how that would help me offhand?